listener/listener.go

```diff
@@ -300,80 +315,150 @@ func (l *Listener) Stream(ctx context.Context) error {
-			return fmt.Errorf("process message: %w", err)
+	topicSet := make(map[string]bool)
+	defer func() {
+		for topic := range topicSet {
+			l.publisher.Flush(topic)
 		}
+	}()
 
-		l.processHeartBeat(msg)
-	}
-}
+	tx := NewWalTransaction(l.log, pool, l.monitor)
 
-func (l *Listener) processMessage(ctx context.Context, msg *pgx.ReplicationMessage, tx *WalTransaction) error {
-	if msg.WalMessage == nil {
-		l.log.Debug("empty wal-message")
-		return nil
-	}
+	group, ctx := errgroup.WithContext(ctx)
+	messageChan := make(chan *pgx.ReplicationMessage, 100)
+	eventsChan := make(chan *messageAndEvents, 100)
+	resultChan := make(chan *eventAndPublishResult, 500)
 
-	l.log.Debug("WAL message has been received", slog.Uint64("wal", msg.WalMessage.WalStart))
+	defer close(messageChan)
+	defer close(eventsChan)
+	defer close(resultChan)
 
-	if err := l.parser.ParseWalMessage(msg.WalMessage.WalData, tx); err != nil {
-		l.monitor.IncProblematicEvents(problemKindParse)
-		return fmt.Errorf("parse: %w", err)
-	}
+	group.Go(func() error {
+		for {
+			if err := ctx.Err(); err != nil {
+				l.log.Warn("stream: context canceled", "err", err)
+				return nil
+			}
 
-	if tx.CommitTime != nil {
-		for event := range tx.CreateEventsWithFilter(ctx, l.cfg.Listener.Filter.Tables) {
-			subjectName := event.SubjectName(l.cfg)
+			msg, err := l.replicator.WaitForReplicationMessage(ctx)
+			if err != nil {
+				return fmt.Errorf("wait for replication message: %w", err)
+			}
 
-			if err := l.publisher.Publish(ctx, subjectName, event); err != nil {
-				l.monitor.IncProblematicEvents(problemKindPublish)
-				return fmt.Errorf("publish: %w", err)
+			if msg == nil {
+				continue
 			}
 
-			l.monitor.IncPublishedEvents(subjectName, event.Table)
+			messageChan <- msg
+		}
+	})
 
-			l.log.Info(
-				"event was sent",
-				slog.String("subject", subjectName),
-				slog.String("action", event.Action),
-				slog.String("table", event.Table),
-				slog.Uint64("lsn", l.readLSN()),
-			)
+	group.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				l.log.Warn("stream: context canceled", "err", ctx.Err())
+				return nil
+			case msg := <-messageChan:
+				if err := l.parser.ParseWalMessage(msg.WalMessage.WalData, tx); err != nil {
+					l.monitor.IncProblematicEvents(problemKindParse)
+					return fmt.Errorf("parse: %w", err)
+				}
+
+				if tx.CommitTime != nil {
+					events := tx.CreateEventsWithFilter(ctx, l.cfg.Listener.Filter.Tables)
+
+					eventsChan <- &messageAndEvents{
+						msg,
+						events,
+					}
+					tx.Clear()
+				}
+
+				l.processHeartBeat(msg)
+			}
+		}
+	})
 
-			tx.pool.Put(event)
+	group.Go(func() error {
+		for {
+			select {
+			case <-ctx.Done():
+				l.log.Warn("stream: context canceled", "err", ctx.Err())
+				return nil
+			case object := <-eventsChan:
+				events, msg := object.events, object.msg
+
+				for idx, event := range events {
+					subjectName := event.SubjectName(l.cfg)
+					topicSet[subjectName] = true
```

Comment:
It's ok that the topicSet isn't a `sync.Map` or what have you cause there is only one of these ever touching it right?